benchmark/benchmark.go

```diff
@@ -0,0 +1,263 @@
+	flag.Usage = Usage
+	flag.Parse()
+	osArgs := flag.Args()
+
+	if a.listen {
+		if len(osArgs) < 2 {
+			Usage()
+			exit("")
+		}
+		a.udtAddr = osArgs[0]
+		a.tcpAddr = osArgs[1]
+	} else {
+		if len(osArgs) < 1 {
+			Usage()
+			exit("")
+		}
+		a.remoteAddr = osArgs[0]
+	}
+
+	return a
+}
+
+func main() {
+	args := parseArgs()
+	verbose = args.verbose
+
+	go func() {
+		// wait until we exit.
+		sigc := make(chan os.Signal, 1)
+		signal.Notify(sigc, syscall.SIGABRT)
+		<-sigc
+		panic("ABORT! ABORT! ABORT!")
+	}()
+
+	var err error
+	if args.listen {
+		err = Listen(args.udtAddr, args.tcpAddr, args.bs)
+	} else {
+		var c net.Conn
+		if args.dialUdt {
+			c, err = Dial("udt", args.remoteAddr)
+		} else {
+			c, err = Dial("tcp", args.remoteAddr)
+		}
+		if err != nil {
+			exit("%s", err)
+		}
+		err = benchmark(c, args.bs)
+	}
+
+	if err != nil {
+		exit("%s", err)
+	}
+}
+
+func exit(format string, vals ...interface{}) {
+	if format != "" {
+		fmt.Fprintf(os.Stderr, "benchmark error: "+format+"\n", vals...)
+	}
+	os.Exit(1)
+}
+
+func log(format string, vals ...interface{}) {
+	if verbose {
+		fmt.Fprintf(os.Stderr, "benchmark log: "+format+"\n", vals...)
+	}
+}
+
+// Listen listens and accepts one incoming UDT connection on a given port,
+// and pipes all incoming data to os.Stdout.
+func Listen(udtAddr string, tcpAddr string, bs int64) error {
+	udt, err := udt.Listen("udt", udtAddr)
+	if err != nil {
+		return err
+	}
+	log("udt listening at %s", udt.Addr())
+
+	tcp, err := net.Listen("tcp", tcpAddr)
+	if err != nil {
+		return err
+	}
+	log("tcp listening at %s", tcp.Addr())
+
+	done := make(chan interface{}, 2)
+	run := func(l net.Listener) (err error) {
+		for {
+			var c net.Conn
+			c, err = l.Accept()
+			if err != nil {
+				return
+			}
+			log("accepted connection from %s", c.RemoteAddr())
+
+			go func() {
+				defer c.Close()
+				for {
+					select {
+					case <-done:
+						return
+					default:
+						n, e := io.CopyN(c, c, bs)
+						if e == io.EOF {
+							log("read EOF, close client socket")
+							return
+						}
+						log("Copied back %d bytes", n)
+					}
+				}
+
+				return
+			}()
+		}
+	}
+
+	go run(udt)
+	go run(tcp)
+	// wait until we exit.
+	sigc := make(chan os.Signal, 1)
+	signal.Notify(sigc, syscall.SIGHUP, syscall.SIGINT,
+		syscall.SIGTERM, syscall.SIGQUIT)
+	<-sigc
+	fmt.Println("Quit now")
+	done <- nil
+	return nil
+}
+
+// Dial connects to a remote address and pipes all os.Stdin to the remote end.
+// If localAddr is set, uses it to Dial from.
+func Dial(network, remoteAddr string) (c net.Conn, err error) {
+	log("%s dialing %s", network, remoteAddr)
+	if network == "udt" {
+		c, err = udt.Dial("udt", remoteAddr)
+	} else {
+		c, err = net.Dial("tcp", remoteAddr)
+	}
+	if err != nil {
+		return
+	}
+	log("connected to %s", c.RemoteAddr())
+	return
+
+}
+
+func benchmark(c net.Conn, bs int64) (err error) {
+	defer c.Close()
+	rand, err := os.Open("/dev/zero")
```

Comment:
yeah, instead of using a file, use a rand reader. either the one in `"crypto/rand"` or checkout @dustin's randbo, its a really fast "kinda random" generator